cabby/geo/walk.py

```diff
@@ -457,31 +458,43 @@ def get_pivot_beyond_goal(self,
       return GeoDataFrame(index=[0], columns=route.columns).iloc[0]
 
     street_beyond_osmid = street_beyond_route['osmid'].iloc[0]
-
-    # Change OSMID to key
-    segment_beyond_path = ((last_node_in_route['osmid'] == self.map.edges['u'])
-                & (before_last_node_in_route['osmid'] !=
-                  self.map.edges['v']))
     condition_street_id = self.map.edges['osmid'].apply(
       lambda x: x == street_beyond_osmid)
-    last_line = self.map.edges[condition_street_id
-                & segment_beyond_path]
-
-    if last_line.shape[0] == 0:
-      # Return Empty.
-      return GeoDataFrame(index=[0], columns=route.columns).iloc[0]
-
-    last_line_max = last_line[last_line['length'] ==
-                  last_line['length'].max()].iloc[0]
-
-    poly = last_line_max['geometry'].buffer(0.0001)
+    street_nodes = self.map.edges[condition_street_id]['u'].unique()
+    street_nodes = np.random.choice(street_nodes, MAX_NUM_BEYOND_TRY)
+    for i in range(MAX_NUM_BEYOND_TRY):
+      length = nx.shortest_path_length(self.map.nx_graph, source=street_nodes[i],target=last_node_in_route['osmid'])
+      if length>3 and length<10:
+        continue
+      path = nx.shortest_path(self.map.nx_graph, source=street_nodes[i], target=last_node_in_route['osmid'])
+      path.remove(last_node_in_route['osmid'])
+      if final_node_in_route['osmid'] in path:
+        path.remove(final_node_in_route['osmid'])
+
+      intersections = set(route['osmid']).intersection(path)
+      if len(intersections)<1 and len(path)>2:
+        beyond = self.select_pivot_from_path(route, end_point, path)
+        if beyond is not None:
+          return beyond
+    return GeoDataFrame(index=[0], columns=route.columns).iloc[0]
+
+
+  def select_pivot_from_path(self,
+                        route: GeoDataFrame,
+                        end_point: Dict,
+                        path: list):
+
+    path_nodes = self.map.nodes[self.map.nodes['osmid'].isin(path)]
+    points_route = path_nodes['geometry'].tolist()
+    poly = LineString(points_route).buffer(PIVOT_ALONG_ROUTE_MAX_DIST)
+    route_shape = LineString(route['geometry'].tolist()).buffer(PIVOT_ALONG_ROUTE_MAX_DIST)
+    poly = poly.difference(route_shape)
```

Comment:
Instead of reusing 'poly', please use a new variable name that documents what the new 'poly' is.